File: apps/windows/src/Utilities/CategoryHierarchy.cs
// Helpers for navigating the category hierarchy (ParentId). Since the JSON
// can be hand-edited, dangling parents (treated as roots) and cycles
// (reattached to the root) are both tolerated.
using LexiCall.Desktop.Models;

namespace LexiCall.Desktop.Utilities;

public static class CategoryHierarchy
{
    private static readonly IReadOnlyDictionary<Guid, int> NoOrder = new Dictionary<Guid, int>();

    // Depth-first walk: roots then children recursively, each sibling group
    // sorted via SortSiblings (manual order if any, otherwise alphabetical).
    // The returned Depth lets flat lists indent without a real tree. order
    // comes from CategoryOrderStore (context menu "Monter"/"Descendre" —
    // Move up/down); omitted, the order stays purely alphabetical.
    public static IReadOnlyList<(VocabularyCategory Category, int Depth)> Flatten(
        IReadOnlyCollection<VocabularyCategory> categories,
        IReadOnlyDictionary<Guid, int>? order = null)
    {
        var (roots, childrenByParent) = BuildChildrenLookup(categories, order ?? NoOrder);
        var result = new List<(VocabularyCategory, int)>(categories.Count);
        var visited = new HashSet<Guid>();

        void Visit(VocabularyCategory category, int depth)
        {
            if (!visited.Add(category.Id))
            {
                return;
            }

            result.Add((category, depth));

            if (childrenByParent.TryGetValue(category.Id, out var children))
            {
                foreach (var child in children)
                {
                    Visit(child, depth + 1);
                }
            }
        }

        foreach (var root in roots)
        {
            Visit(root, 0);
        }

        // Categories never reached from a root: a parenting cycle — bump
        // them up to the root so they stay visible.
        foreach (var category in categories.OrderBy(category => category.Name, StringComparer.CurrentCultureIgnoreCase))
        {
            Visit(category, 0);
        }

        return result;
    }

    // Assigns each category its root's color index; descendants inherit
    // their root's index. Backs CategoryColorResolver's automatic fallback
    // when no color was manually chosen (see CategoryColorStore). A root's
    // index is derived from its Id (stable), not its position among other
    // roots: reordering categories (Move up/down) must never change their
    // automatic color.
    public static IReadOnlyDictionary<Guid, int> ComputeColorIndexes(IReadOnlyCollection<VocabularyCategory> categories)
    {
        var colorIndexes = new Dictionary<Guid, int>();
        var lastIndexByDepth = new List<int>();

        foreach (var (category, depth) in Flatten(categories))
        {
            var colorIndex = depth == 0 ? GetStableIndex(category.Id) : lastIndexByDepth[depth - 1];

            if (lastIndexByDepth.Count > depth)
            {
                lastIndexByDepth[depth] = colorIndex;
                lastIndexByDepth.RemoveRange(depth + 1, lastIndexByDepth.Count - depth - 1);
            }
            else
            {
                lastIndexByDepth.Add(colorIndex);
            }

            colorIndexes[category.Id] = colorIndex;
        }

        return colorIndexes;
    }

    // & int.MaxValue rather than Math.Abs: avoids the OverflowException on
    // int.MinValue (Math.Abs(int.MinValue) has no positive representation).
    private static int GetStableIndex(Guid categoryId) => categoryId.GetHashCode() & int.MaxValue;

    public static HashSet<Guid> GetDescendantIds(
        IReadOnlyCollection<VocabularyCategory> categories,
        Guid rootId)
    {
        var (_, childrenByParent) = BuildChildrenLookup(categories, NoOrder);
        var descendants = new HashSet<Guid>();

        void Visit(Guid parentId)
        {
            if (!childrenByParent.TryGetValue(parentId, out var children))
            {
                return;
            }

            foreach (var child in children)
            {
                if (descendants.Add(child.Id))
                {
                    Visit(child.Id);
                }
            }
        }

        Visit(rootId);
        return descendants;
    }

    // Siblings (same effective parent as category, itself included) in the
    // current display order — used to move a category up/down among its
    // siblings (MainWindowViewModel.MoveCategoryUp/Down).
    public static IReadOnlyList<VocabularyCategory> GetSiblingsInOrder(
        IReadOnlyCollection<VocabularyCategory> categories,
        VocabularyCategory category,
        IReadOnlyDictionary<Guid, int>? order = null)
    {
        var knownIds = categories.Select(c => c.Id).ToHashSet();
        var parentId = EffectiveParentId(category, knownIds);
        var siblings = categories.Where(c => EffectiveParentId(c, knownIds) == parentId).ToList();
        SortSiblings(siblings, order ?? NoOrder);
        return siblings;
    }

    private static (List<VocabularyCategory> Roots, Dictionary<Guid, List<VocabularyCategory>> ChildrenByParent)
        BuildChildrenLookup(IReadOnlyCollection<VocabularyCategory> categories, IReadOnlyDictionary<Guid, int> order)
    {
        var knownIds = categories.Select(category => category.Id).ToHashSet();
        var roots = new List<VocabularyCategory>();
        var childrenByParent = new Dictionary<Guid, List<VocabularyCategory>>();

        foreach (var category in categories)
        {
            if (EffectiveParentId(category, knownIds) is Guid parentId)
            {
                if (!childrenByParent.TryGetValue(parentId, out var children))
                {
                    children = [];
                    childrenByParent[parentId] = children;
                }

                children.Add(category);
            }
            else
            {
                roots.Add(category);
            }
        }

        SortSiblings(roots, order);

        foreach (var children in childrenByParent.Values)
        {
            SortSiblings(children, order);
        }

        return (roots, childrenByParent);
    }

    // Manual rank (CategoryOrderStore) takes priority if either compared
    // category has one, otherwise alphabetical — a category added later to
    // an already-reordered group sorts alphabetically after the manually
    // positioned ones, until it's moved in turn.
    private static void SortSiblings(List<VocabularyCategory> siblings, IReadOnlyDictionary<Guid, int> order)
    {
        siblings.Sort((a, b) =>
        {
            var hasOrderA = order.TryGetValue(a.Id, out var rankA);
            var hasOrderB = order.TryGetValue(b.Id, out var rankB);

            if (hasOrderA && hasOrderB)
            {
                return rankA.CompareTo(rankB);
            }

            if (hasOrderA != hasOrderB)
            {
                return hasOrderA ? -1 : 1;
            }

            return string.Compare(a.Name, b.Name, StringComparison.CurrentCultureIgnoreCase);
        });
    }

    private static Guid? EffectiveParentId(VocabularyCategory category, HashSet<Guid> knownIds)
    {
        return category.ParentId is Guid parentId && knownIds.Contains(parentId) && parentId != category.Id
            ? parentId
            : null;
    }
}

File: apps/android/src/utils/categoryHierarchy.js
// Navigating the category hierarchy (ParentId). A dangling parent is treated as
// a root and a parenting cycle is reattached to the root, so hand-edited or
// partially-synced data still renders.
import { colorIndexForRoot } from './categoryColor';
import { compareText } from './collation';

function effectiveParentId(category, knownIds) {
  const parentId = category.ParentId;
  return parentId && parentId !== category.Id && knownIds.has(parentId) ? parentId : null;
}

function compareNames(a, b) {
  return compareText(a.Name, b.Name);
}

// Manual rank (from settings.json, see the store's categoryOrder) takes
// priority if either compared category has one, otherwise alphabetical — a
// category added later to an already-reordered group sorts alphabetically
// after the manually positioned ones, until it's moved in turn.
function sortSiblings(siblings, order) {
  siblings.sort((a, b) => {
    const rankA = order[a.Id];
    const rankB = order[b.Id];
    const hasRankA = rankA !== undefined;
    const hasRankB = rankB !== undefined;

    if (hasRankA && hasRankB) {
      return rankA - rankB;
    }
    if (hasRankA !== hasRankB) {
      return hasRankA ? -1 : 1;
    }
    return compareNames(a, b);
  });
}

function buildChildrenLookup(categories, order = {}) {
  const knownIds = new Set(categories.map((category) => category.Id));
  const roots = [];
  const childrenByParent = new Map();

  for (const category of categories) {
    const parentId = effectiveParentId(category, knownIds);

    if (parentId === null) {
      roots.push(category);
      continue;
    }

    const siblings = childrenByParent.get(parentId) ?? [];
    siblings.push(category);
    childrenByParent.set(parentId, siblings);
  }

  sortSiblings(roots, order);
  for (const siblings of childrenByParent.values()) {
    sortSiblings(siblings, order);
  }

  return { roots, childrenByParent };
}

// Depth-first walk returning [{ category, depth }], roots first then children.
// order (categoryOrder from settings) breaks ties within a sibling group;
// omitted, siblings stay purely alphabetical.
export function flattenCategories(categories, order = {}) {
  const { roots, childrenByParent } = buildChildrenLookup(categories, order);
  const flattened = [];
  const visited = new Set();

  const visit = (category, depth) => {
    if (visited.has(category.Id)) {
      return;
    }
    visited.add(category.Id);
    flattened.push({ category, depth });

    for (const child of childrenByParent.get(category.Id) ?? []) {
      visit(child, depth + 1);
    }
  };

  for (const root of roots) {
    visit(root, 0);
  }
  // Anything a root never reached sits in a cycle; surface it at root level
  // rather than letting it vanish from the tree.
  for (const category of [...categories].sort(compareNames)) {
    visit(category, 0);
  }

  return flattened;
}

// Siblings (same effective parent as category, itself included) in the
// current display order — used to move a category up/down among them.
export function getSiblingsInOrder(categories, category, order = {}) {
  const knownIds = new Set(categories.map((item) => item.Id));
  const parentId = effectiveParentId(category, knownIds);
  const siblings = categories.filter((item) => effectiveParentId(item, knownIds) === parentId);
  sortSiblings(siblings, order);
  return siblings;
}

export function getDescendantIds(categories, rootId) {
  const { childrenByParent } = buildChildrenLookup(categories);
  const descendants = new Set();

  const visit = (parentId) => {
    for (const child of childrenByParent.get(parentId) ?? []) {
      if (!descendants.has(child.Id)) {
        descendants.add(child.Id);
        visit(child.Id);
      }
    }
  };

  visit(rootId);
  return descendants;
}

// Each category inherits its root's color index, so a subtree reads as one
// color family. The index comes from the root's own Id rather than its position
// among siblings: reordering categories must never repaint them.
export function computeColorIndexes(categories) {
  const indexes = new Map();
  const lastIndexByDepth = [];

  for (const { category, depth } of flattenCategories(categories)) {
    const colorIndex = depth === 0 ? colorIndexForRoot(category.Id) : lastIndexByDepth[depth - 1];

    lastIndexByDepth[depth] = colorIndex;
    lastIndexByDepth.length = depth + 1;
    indexes.set(category.Id, colorIndex);
  }

  return indexes;
}
